workers/data_refinery_workers/downloaders/array_express.py

```diff
@@ -33,20 +30,16 @@ def download_array_express(job_id):
         utils.end_job(job, None, False)
         return
 
-    target_directory = ROOT_URI + batch.internal_location
-    os.makedirs(target_directory, exist_ok=True)
-
-    filename = batch.download_url.split('/')[-1]
-    target_file_name = target_directory + filename
+    target_file_path = utils.prepare_destination(batch)
 
     logger.info("Downloading file from %s to %s. (Batch #%d, Job #%d)",
                 batch.download_url,
-                target_file_name,
+                target_file_path,
                 batch.id,
                 job_id)
 
     try:
-        target_file = open(target_file_name, "wb")
+        target_file = open(target_file_path, "wb")
         request = requests.get(batch.download_url, stream=True)
 
         for chunk in request.iter_content(CHUNK_SIZE):
```

Comment:
TBH I didn't change this download code too much from what it was when [Rene wrote it](https://bitbucket.org/greenelab/arrayprocess/src/a34824df6100d8bd999b99cf4cab03d502b985ee/download_raw_data_files_by_platform.py?at=default&fileviewer=file-view-default). Now that I've put some thought into this I'd agree this `CHUNK_SIZE` is rather small. However what we actually want it to be will probably depend on the amount of memory on the machines we run it on. I think a quarter of a GB might be a good size for now though. It's large enough to get some performance gains but even if four of them are running in sync the memory usage won't be much over a GB.

And I agree that could be a good future feature. However it seems likely that downloading is going to take much less time than processing so unless this becomes an issue where we have trouble keeping a connection open long enough to download large files I think that this optimization would have a fairly small relative effect. Definitely something to keep in mind for the future though.